src/librustc/ty/sty.rs

```diff
@@ -571,14 +571,28 @@ impl<'a, 'tcx> ProjectionTy<'tcx> {
             |item| item.name == item_name).unwrap().def_id;
 
         ProjectionTy {
-            trait_ref: trait_ref,
+            substs: trait_ref.substs,
             item_def_id: item_def_id,
         }
     }
 
-    pub fn item_name(self, tcx: TyCtxt) -> Name {
+    pub fn item_name(&self, tcx: TyCtxt) -> Name {
+        // FIXME(tschottdorf): scrutinize call sites. Should only be used
```

Comment:
A lazy cursory search (grep for `proj.*item_name`) suggests that the only such caller is to supply a `infer::LateBoundRegionConversionTime::AssocTypeProjection(item_name)`. I'll take a look at at least not passing `item_name` in...